src/app/inventory/store/sockets.ts

```diff
@@ -409,6 +416,92 @@ function addPlugOption(
+ *    slots don't reference any plug in singleInitialItemHash, and the first
+ *    entry in the plug set just so happened to be the empty plug.
+ * 2. The socket's singleInitialItemHash may reference a non-empty plug. A lot
+ *    of armor references associated shaders here instead.
+ * 3. The singleInitialItemHash is an empty plug, but it's not the proper empty
+ *    plug. This happens to void subclass aspect and fragment sockets, and is
+ *    really insidious because void subclasses start with these sockets but these
+ *    plugs can never be inserted, so we can't use it.
+ */
+function findEmptyPlug(
+  socket: DestinyItemSocketEntryDefinition,
+  socketType: DestinySocketTypeDefinition,
+  plugSet: DimPlugSet | undefined,
+  reusablePlugs?: DestinyItemPlugBase[]
+) {
+  // First, perform some filtering so we don't repeatedly search through very large plug sets
+  // like armor 2.0 stat plug sets, masterwork sets, or armor energy sets.
+
+  if (noDefaultSocketCategoryHashes.includes(socketType.socketCategoryHash)) {
+    return undefined;
+  }
+
+  // Y2+ weapon masterworks don't have an "empty" entry.
+  if (socket.socketTypeHash === weaponMasterworkY2SocketTypeHash) {
+    return undefined;
+  }
+  // Exotic mods (like the Aeon socket) can't be emptied.
+  if (
+    socketType.plugWhitelist.length &&
+    socketType.plugWhitelist.every((e) => e.categoryIdentifier.includes('enhancements.exotic'))
+  ) {
+    return undefined;
+  }
+  // Sockets that ONLY get their items from your inventory necessarily can't be emptied
+  if ((socket.plugSources & ~SocketPlugSources.InventorySourced) === 0) {
+    return undefined;
+  }
+
+  // Check sources in decreasing order of assumed reliability:
+  // When the live API response tells us about a certain empty plug, that's
+  // probably the most correct. PlugSets are usually better than the
+  // socket.reusablePlugs, but sometimes there's an empty option not present in the PlugSet.
+  // FIXME #7793: reusablePlugItems is thrown away when there's a PlugSet.
+  const isDefault = (p: number) => emptyPlugHashes.has(p);
+  const empty =
+    reusablePlugs?.map((p) => p.plugItemHash).find(isDefault) ||
+    plugSet?.plugs.map((p) => p.plugDef.hash).find(isDefault) ||
+    socket.reusablePlugItems.map((p) => p.plugItemHash).find(isDefault);
+
+  // Falling back to singleInitialItemHash is the conservative choice:
+  // 1. Before this function existed, we used singleInitialItemHash all the
+  //    time and it only broke in specific situations, so we might as well
+  //    continue using it when we didn't find a better plug before.
+  // 2. The game has a lot of sockets and we don't want to be updating
+  //    D2AI every time a new socket appears -- better to just fix either
+  //    the filters above or the D2AI list when something breaks.
+  //
+  // If there's a very good reason to assume a socket can't be emptied, filter it above.
+  return empty ? empty : socket.singleInitialItemHash || undefined;
```

Comment:
`empty ?? socket.singleInitialItemHash || undefined` maybe?
depends a little on whether you consider this more of an if/then or a series of 2 fallbacks